engine/consensus/sealing/engine.go

```diff
@@ -23,240 +22,280 @@ type Event struct {
 	}
-
-	// register engine with the receipt provider
-	_, err = net.Register(engine.ReceiveReceipts, e)
-	if err != nil {
-		return nil, fmt.Errorf("could not register for results: %w", err)
+	e.pendingRequestedApprovals = &engine.FifoMessageStore{
+		FifoQueue: pendingRequestedApprovalsQueue,
 	}
 
-	// register engine with the approval provider
-	_, err = net.Register(engine.ReceiveApprovals, e)
+	e.pendingFinalizationEvents, err = fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(defaultFinalizationEventsQueueCapacity))
 	if err != nil {
-		return nil, fmt.Errorf("could not register for approvals: %w", err)
+		return fmt.Errorf("failed to create queue for finalization events: %w", err)
 	}
 
-	// register engine to the channel for requesting missing approvals
-	approvalConduit, err := net.Register(engine.RequestApprovalsByChunk, e)
+	e.pendingIncorporatedResults, err = fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(defaultIncorporatedResultsQueueCapacity))
 	if err != nil {
-		return nil, fmt.Errorf("could not register for requesting approvals: %w", err)
+		return fmt.Errorf("failed to create queue for incorproated results: %w", err)
 	}
 
-	e.core, err = NewCore(log, engineMetrics, tracer, mempool, conMetrics, state, me, receiptRequester, receiptsDB, headersDB,
-		indexDB, incorporatedResults, receipts, approvals, seals, pendingReceipts, assigner, receiptValidator, approvalValidator,
-		requiredApprovalsForSealConstruction, emergencySealingActive, approvalConduit)
-	if err != nil {
-		return nil, fmt.Errorf("failed to init sealing engine: %w", err)
-	}
+	e.notifier = engine.NewNotifier()
+	// define message queueing behaviour
+	e.messageHandler = engine.NewMessageHandler(
+		e.log,
+		e.notifier,
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				_, ok := msg.Payload.(*flow.ResultApproval)
+				if ok {
+					e.engineMetrics.MessageReceived(metrics.EngineSealing, metrics.MessageResultApproval)
+				}
+				return ok
+			},
+			Map: func(msg *engine.Message) (*engine.Message, bool) {
+				if e.requiredApprovalsForSealConstruction < 1 {
+					// if we don't require approvals to construct a seal, don't even process approvals.
+					return nil, false
+				}
 
-	return e, nil
+				return msg, true
+			},
+			Store: e.pendingApprovals,
+		},
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				_, ok := msg.Payload.(*messages.ApprovalResponse)
+				if ok {
+					e.engineMetrics.MessageReceived(metrics.EngineSealing, metrics.MessageResultApproval)
+				}
+				return ok
+			},
+			Map: func(msg *engine.Message) (*engine.Message, bool) {
+				if e.requiredApprovalsForSealConstruction < 1 {
+					// if we don't require approvals to construct a seal, don't even process approvals.
+					return nil, false
+				}
+
+				return &engine.Message{
+					OriginID: msg.OriginID,
+					Payload:  msg.Payload.(*messages.ApprovalResponse).Approval,
+				}, true
```

Comment:
```suggestion
				approval := msg.Payload.(*messages.ApprovalResponse).Approval
				return &approval,
				}, true
```
⚠️ 
The `Approval` field is a struct, but here we need it to be a pointer, because the consumer will [convert convert the payload into a `*flow.ResultApproval`](https://github.com/onflow/flow-go/pull/736/files#diff-bb47ed18fad3632a211e4d6230f9b0b17f42693c309b01f8f96d82ca71c6ffc2R249) (L249).

Better to have a test case to verify this